### DISCUSSION TITLE
Use `pydantic-settings` to better manage environment variables

### DISCUSSION THREAD

**Discussion by gbaian10** [JoshuaC215/agent-service-toolkit#64] [Ideas]
[pydantic-settings document](https://docs.pydantic.dev/latest/concepts/pydantic_settings/)

I hope to incorporate pydantic-settings to manage environment variables uniformly.
Each time we use environment variables, we should access them through this object instead of always using `os.getenv` or `os.environ`.

If you think this is a good idea, I can provide a PR to accomplish it.

**Comment by JoshuaC215** [comment]
Nice! I wasn't familiar with this package but it sounds great, will solve a few problems. Doesn't look like it's super popular but obviously pydantic support goes a long way.

I did a little research and found e.g. this: https://www.reddit.com/r/learnpython/comments/12cm6cm/library_for_python_similar_to_gos_viper_12_factor/

Seems like there isn't another great and currently maintained option for this problem.

Let's do it! Thanks!

**Comment by gbaian10** [comment]
FastAPI template also uses it, which can refer to

https://github.com/fastapi/full-stack-fastapi-template/blob/d3d370cad04228a9e5a4a374a007c0f4f9f068a7/backend/app/core/config.py#L26

**Comment by JoshuaC215** [comment]
I was trying to figure out how to do value patching in tests and found a couple of references - putting here to have handy later if needed

https://rednafi.com/python/patch_pydantic_settings_in_pytest/
https://github.com/na86421/override-pydantic-settings/blob/master/override_settings/override.py#L35

**Comment by gbaian10** [comment]
Have you considered making an LLM from the same provider configurable?
For example, choosing between OpenAI's 4o-mini or 4o, or selecting Anthropic's Haiku or Sonnet.

This change will primarily affect the API input (as well as the input options for Streamlit).

**Reply by JoshuaC215** [reply]
I thought about it. Since this is intended as a template I’m not sure how useful it is vs just changing to the desired model set after you clone.

What do you think? If there’s an easy change, not opposed to it but don’t want to get into maintaining some huge list or doing a bunch of error parsing to clean up invalid entries etc

**Reply by gbaian10** [reply]
I have a draft that involves quite a bit of code modification, but future maintenance will be much easier.
Most of it is accomplished using Enums, and ultimately you only need to maintain this model list in one file to achieve validation and API docs.

**Reply by JoshuaC215** [reply]
👍 Feel free to post and we can take a look and decide
